source/common/io/io_uring_worker_impl.cc

```diff
@@ -123,5 +275,340 @@ void IoUringWorkerImpl::submit() {
+      return;
+    }
+  }
+
+  // Move read data from request to buffer or store the error.
+  if (result > 0) {
+    moveReadDataToBuffer(req, result);
+  } else {
+    if (result != -ECANCELED) {
+      read_error_ = result;
+    }
+  }
+
+  // Discard calling back since the socket is not ready or closed.
+  if (status_ == Initialized || status_ == Closed) {
+    return;
+  }
+
+  // If the socket is enabled and there are bytes to read, notify the handler.
+  if (status_ == ReadEnabled) {
+    if (read_buf_.length() > 0) {
+      onReadCompleted(static_cast<int32_t>(read_buf_.length()));
+    } else if (read_error_.has_value() && read_error_ < 0) {
+      onReadCompleted(read_error_.value());
+      read_error_.reset();
+    }
+    // Handle remote closed at last.
+    // Depending on the event listened to, calling different event back to handle remote closed.
+    // * events & (Read | Closed): Callback Closed,
+    // * events & (Read)         : Callback Read,
+    // * events & (Closed)       : Callback Closed,
+    // * ...else                 : Callback Write.
+    if (read_error_.has_value() && read_error_ == 0 && !enable_close_event_) {
+      ENVOY_LOG(trace, "read remote closed from socket, fd = {}", fd_);
+      onReadCompleted(read_error_.value());
+      read_error_.reset();
+      return;
+    }
+  }
+
+  // If `enable_close_event_` is true, then deliver the remote close as close event.
+  if (read_error_.has_value() && read_error_ == 0) {
+    if (enable_close_event_) {
+      ENVOY_LOG(trace,
+                "remote closed and close event enabled, raise the close event, fd = "
+                "{}, result = {}",
+                fd_, read_error_.value());
+      status_ = RemoteClosed;
+      IoUringSocketEntry::onRemoteClose();
+      read_error_.reset();
+      return;
+    } else {
+      // In this case, the closed event isn't listened and the status is disabled.
+      // It means we can't raise the closed or read event. So we only can raise the
+      // write event.
+      ENVOY_LOG(trace,
+                "remote closed and close event disabled, raise the write event, fd = "
+                "{}, result = {}",
+                fd_, read_error_.value());
+      status_ = RemoteClosed;
+      onWriteCompleted(0);
+      read_error_.reset();
+      return;
+    }
+  }
+
+  // The socket may be not readable during handler onRead callback, check it again here.
+  if (status_ == ReadEnabled) {
+    // If the read error is zero, it means remote close, then needn't new request.
+    if (!read_error_.has_value() || read_error_.value() != 0) {
+      // Submit a read accept request for the next read.
+      submitReadRequest();
+    }
+  } else if (status_ == ReadDisabled) {
+    // Since error in a disabled socket will not be handled by the handler, stop submit read
+    // request if there is any error.
+    if (!read_error_.has_value()) {
+      // Submit a read accept request for the next read.
+      submitReadRequest();
```

Comment:
In the read disabled state, why is the next read request needed to be submitted? I would think it should be submitted when the read is enabled.

Reply:
This is for the monitor the remote close of the connection, otherwise, we have no way to know the client closed the connection.